scheduler/scheduler.cpp

```diff
@@ -624,72 +626,104 @@ static string envs_match(CompileServer *cs, const Job *job)
     return string();
 }
 
-static CompileServer *pick_server(Job *job)
+static list<CompileServer *> filter_ineligible_servers(Job *job)
 {
+    list<CompileServer *> eligible;
+    std::copy_if(
+        eligible.begin(),
+        eligible.end(),
+        std::back_inserter(eligible),
+        [=](CompileServer* cs) {
+            if (!cs->is_eligible_now(job)) {
 #if DEBUG_SCHEDULER > 1
-    trace() << "pick_server " << job->id() << " " << job->targetPlatform() << endl;
+                if ((int(cs->jobList().size()) >= cs->maxJobs() + cs->maxPreloadCount()) || (cs->load() >= 1000)) {
+                    trace() << "overloaded " << cs->nodeName() << " " << cs->jobList().size() << "/"
+                            <<  cs->maxJobs() << " jobs, load:" << cs->load() << endl;
+                } else
+                    trace() << cs->nodeName() << " not eligible" << endl;
 #endif
+                return false;
+            }
 
-#if DEBUG_SCHEDULER > 0
+            // incompatible architecture or busy installing
+            if (!cs->can_install(job).size()) {
+#if DEBUG_SCHEDULER > 2
+                trace() << cs->nodeName() << " can't install " << job->id() << endl;
+#endif
+                return false;
+            }
 
-    /* consistency checking for now */
-    for (list<CompileServer *>::iterator it = css.begin(); it != css.end(); ++it) {
-        CompileServer *cs = *it;
+            /* Don't use non-chroot-able daemons for remote jobs.  XXX */
+            if (!cs->chrootPossible() && cs != job->submitter()) {
+                trace() << cs->nodeName() << " can't use chroot\n";
+                return false;
+            }
 
-        list<Job *> jobList = cs->jobList();
-        for (list<Job *>::const_iterator it2 = jobList.begin(); it2 != jobList.end(); ++it2) {
-            assert(jobs.find((*it2)->id()) != jobs.end());
-        }
-    }
+            // Check if remote & if remote allowed
+            if (!cs->check_remote(job)) {
+                trace() << cs->nodeName() << " fails remote job check\n";
```

Comment:
debug define?

Reply:
Nope. `trace` is just the debug log stream. The only time they put anything behind a debug define is when they're doing something expensive.

Reply:
Which is as it should be - if there are problems it is a lot better if we can say enable logs and send them to us vs rebuild in debug first.